File: 416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = Arrays.stream(nums).sum();
        if(sum%2 != 0) return false;
        return solve(nums , sum/2);
    }

    private boolean solve(int[] arr , int sum){
        int n = arr.length;
        boolean[][] dp = new boolean[n][sum + 1];

        for(int i = 0 ; i < n ; i++){
            dp[i][0] = true;
        }

        if(arr[0] <= sum){
            dp[0][arr[0]] = true;
        }

        for(int i = 1 ; i < n ; i++){
            for(int j = 1 ; j <= sum ; j++){
                boolean np = dp[i-1][j];
                boolean p = false;
                if(arr[i] <= j){
                    p = dp[i-1][j - arr[i]];
                }

                dp[i][j] = np || p;
            }
        }

        return dp[n-1][sum];
    }
}

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        s = sum(nums)

        if s % 2 != 0:
            return False

        n = len(nums)
        target = s // 2

        dp = [[False for _ in range(target + 1)] for _ in range(n)]

        for t in range(target+1):
            dp[0][t] = t == nums[0]

        for i in range(1, n):
            for t in range(target+1):
                np = dp[i - 1][t]
                p = False
                if t >= nums[i]:
                    p = dp[i - 1][t - nums[i]]

                dp[i][t] = p | np

        return dp[n - 1][target]
